### DISCUSSION TITLE
Automatically set up Node.js interpreter for ddev projects

### DISCUSSION THREAD

**Discussion by mikesnoeren** [php-perfect/ddev-intellij-plugin#155] [Ideas]
It would be great if interpreters like Node.js automatically get set-up so we have tighter integration between PhpStorm and DDEV. Currently I still have to set this up manually (or, maybe I am doing something wrong?)

**Comment by rfay** [comment]
I know you know that DDEV already has node set up; what is it you'd like out of this plugin? https://ddev.readthedocs.io/en/latest/users/extend/customization-extendibility/#using-nodejs-with-ddev

**Reply by mikesnoeren** [reply]
Hi @rfay, nice to see you here.

My request is simply to have this plugin set-up the Node.js interpreter in PHPstorm automatically.
I know I can set this up manually, but as this plugin sets up a lot already I thought this would make a nice addition.
![image](https://user-images.githubusercontent.com/52166423/216993214-4bceb934-2a24-4a8a-a342-e07afd892165.png)

**Reply by rfay** [reply]
So you're looking for the node binary inside the container to be available to PhpStorm? 

Have you figured out how to do it manually? It seems you can do it with PhpStorm the same as we do with php as remote interpreter:

<img width="783" alt="Cursor_and_Configure_Node_js_Remote_Interpreter_and_Node_js_Interpreters_and_Settings" src="https://user-images.githubusercontent.com/112444/216994618-5b274385-f59b-42fe-8832-121bb28533f5.png">

So that's the first step, and maybe you would want to do a docs PR explaining it.

However, this isn't actual "integration", just providing the node and yarn tools.

**Reply by mikesnoeren** [reply]
Yes that's exactly what I meant.

I tried manually adding it, but couldn't get it to work properly.

- The option `Coding assistance for Node.js` is always grayed out when I use the Node interpreter in ddev.
- I am missing version numbers, like visible in your screenshot.
- My tailwindCSS auto completion doesn't work.

When installing Node.js locally, all of the above works fine.

Here are my configurations for ddev (and of course, I also have the ddev intellij plugin installed):
![node interpreter settings](https://user-images.githubusercontent.com/52166423/217001837-d2e084c0-8a8a-428e-9f4c-de56af8a39ab.png)
![docker-compose settings](https://user-images.githubusercontent.com/52166423/217002404-43d90d81-55a8-404a-8a26-4d01f7b46792.png)

**Reply by rfay** [reply]
Yeah, so this is territory that doesn't have to do with the plugin at this point. You'll first have to figure out how to get PhpStorm to do things, then get Jetbrains to fix the things that you can't make work, using simple requests in their issue queue, https://youtrack.jetbrains.com/issues/WI

**Reply by mikesnoeren** [reply]
Thanks, I have created an issue in the Jetbrains issue tracker.

However, even when this is resolved on my end, I think having this plugin set this up automatically for every ddev project is a great thing.

**Reply by rfay** [reply]
Please link to the issue, thanks. I agree it will be great to have the plugin set it up... when it's possible to set it up.

**Reply by mikesnoeren** [reply]
Sure, here is the issue: https://youtrack.jetbrains.com/issue/WI-71201/Issues-adding-Docker-colima-Node-interpreter

**Reply by rfay** [reply]
I added it to my watch list, but I would predict it's mostly too complicated for them to act on. To get them to act, you'll want to give them a simple docker-compose-only technique. Make a simple docker-compose.yaml which runs a simple container with node in it, and demonstrate how you can't do what you want to do there.  They think of the world in the basic parts, and don't think about DDEV, etc.

**Reply by mikesnoeren** [reply]
Makes sense! 

For now, I'll await their reaction and if they require more information I will do as you suggested. 
For the time being I will either keep trying to get it to work, or install node locally.. 

I am still wondering if it is an issue with PhpStorm or if I have something misconfigured.
The only reason I need Node.js available in PhpStorm is for the [Tailwind CSS](https://plugins.jetbrains.com/plugin/15321-tailwind-css) plugin.

**Comment by nico-loeber** [comment]
Hey @mikesnoeren,

Thanks for this suggestion, I really like the idea!

The first Problem I do see is that PHPStorm / WebStorm does only support local and WSL installations of the node binary by default.
To configure a nodejs binary in a docker compose environment you need the plugin [Node.js Remote Interpreter](https://plugins.jetbrains.com/plugin/8116-node-js-remote-interpreter).
It's possible to have plugin dependencies on other plugins and they may even be optional, but it makes things a bit more complicated to implement, instead of a bundled plugin that is always available.

The second problem you talked about with @rfay already. First we need to make sure the integration is working when manually configuring it.
And now some good news, for me it is working when entering the full path to the binary `/usr/bin/node` to the configuration! :)
![image](https://user-images.githubusercontent.com/18308418/217037489-dacec9f6-8dcf-4e9f-9175-3463f76b45d7.png)

At least npm is detected in the right version, so that looks all good to me.

I guess the problem is, that /usr/bin/ is not in the PATH of the ddev container for a non login shell that phpstorm is using or something like that. But entering the full path is no problem at all for the plugin.

I will convert this to an issue so we can implement it.

**Reply by mikesnoeren** [reply]
Hi @nico-loeber, 

Thank you for taking your time to help me out.

Correct me if I am wrong, but Isn't the [Node.js Remote Interpreter](https://plugins.jetbrains.com/plugin/8116-node-js-remote-interpreter) a default plugin? I can't recall installing it.

Anyways, I do not notice any difference between setting or not setting the full Node.js interpreter path, in both cases I see a version number for NPM, but never for Node.js. And in both cases the checkmark `Coding assistance for Node.js` is grayed out.. When installing Node.js locally, this checkmark is the one that makes the TailwindCSS plugin work. Does that checkmark work for you? (just to verify there is a problem on my end)

Good to see as an issue for v1.1!

**Comment by nico-loeber** [comment]
* https://github.com/php-perfect/ddev-intellij-plugin/issues/156
